File: python/TEST_syscall_linux.py
#!/usr/bin/env python
# coding: utf-8
# vim: set sw=4 ts=4 sts=4 et tw=80 :

import errno, os, signal, sys, unittest

sys.path.insert(0, '.')
from pinktrace import bitness, event, syscall, trace

class TestSyscallLinux_01(unittest.TestCase):

    def test_01_get_no(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            os.kill(os.getpid(), 0)
            os._exit(0)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid, trace.OPTION_SYSGOOD)

            # Loop until we get to the kill() system call as there's no
            # guarantee that other system calls won't be called beforehand.
            ev = -1
            found = False
            while ev != event.EVENT_EXIT_GENUINE:
                trace.syscall(pid)
                pid, status = os.waitpid(pid, 0)

                ev = event.decide(status)
                if ev == event.EVENT_SYSCALL:
                    scno = syscall.get_no(pid)
                    name = syscall.name(scno)
                    if name == 'kill':
                        found = True
                        break

            self.assert_(found)

            try: trace.kill(pid)
            except OSError: pass

    def test_02_set_no(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            os.kill(os.getpid(), 0)
            os._exit(0)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid, trace.OPTION_SYSGOOD)

            # Loop until we get to the kill() system call as there's no
            # guarantee that other system calls won't be called beforehand.
            ev = -1
            while ev != event.EVENT_EXIT_GENUINE:
                trace.syscall(pid)
                pid, status = os.waitpid(pid, 0)

                ev = event.decide(status)
                if ev == event.EVENT_SYSCALL:
                    scno = syscall.get_no(pid)
                    name = syscall.name(scno)
                    if name == 'kill':
                        syscall.set_no(pid, syscall.INVALID)
                        scno = syscall.get_no(pid)
                        self.assertEqual(scno, syscall.INVALID)
                        break

            try: trace.kill(pid)
            except OSError: pass

    def test_03_get_ret_success(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            os.kill(os.getpid(), 0)
            os._exit(0)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid, trace.OPTION_SYSGOOD)

            # Loop until we get to the kill() system call as there's no
            # guarantee that other system calls won't be called beforehand.
            ev = -1
            insyscall = False
            while ev != event.EVENT_EXIT_GENUINE:
                trace.syscall(pid)
                pid, status = os.waitpid(pid, 0)

                ev = event.decide(status)
                if ev == event.EVENT_SYSCALL:
                    scno = syscall.get_no(pid)
                    name = syscall.name(scno)
                    if insyscall and name == 'kill':
                        ret = syscall.get_ret(pid)
                        self.assertEqual(ret, 0)

                if not insyscall:
                    insyscall = True
                else:
                    insyscall = False

    def test_04_get_ret_fail(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            try:
                open('')
            except IOError:
                pass
            os._exit(0)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid, trace.OPTION_SYSGOOD)

            # Loop until we get to the open() system call as there's no
            # guarantee that other system calls won't be called beforehand.
            ev = -1
            insyscall = False
            while ev != event.EVENT_EXIT_GENUINE:
                trace.syscall(pid)
                pid, status = os.waitpid(pid, 0)

                ev = event.decide(status)
                if ev == event.EVENT_SYSCALL:
                    scno = syscall.get_no(pid)
                    name = syscall.name(scno)
                    if insyscall and name == 'open':
                        ret = syscall.get_ret(pid)
                        self.assertEqual(ret, -errno.ENOENT)

                if not insyscall:
                    insyscall = True
                else:
                    insyscall = False

    def test_05_set_ret_success(self):
        # TODO: Find a smart test case for this one.
        pass

    def test_06_set_ret_fail(self):
        pid = os.fork()
        if not pid: # child
            trace.me()
            os.kill(os.getpid(), signal.SIGSTOP)

            try:
                os.kill(os.getpid(), 0)
            except OSError:
                os._exit(0)
            else:
                os._exit(1)
        else: # parent
            os.waitpid(pid, 0)
            trace.setup(pid, trace.OPTION_SYSGOOD)

            # Loop until we get to the kill() system call as there's no
            # guarantee that other system calls won't be called beforehand.
            ev = -1
            insyscall = False
            while ev != event.EVENT_EXIT_GENUINE:
                trace.syscall(pid)
                pid, status = os.waitpid(pid, 0)

                ev = event.decide(status)
                if ev == event.EVENT_SYSCALL:
                    scno = syscall.get_no(pid)
                    name = syscall.name(scno)
                    if insyscall and name == 'kill':
                        syscall.set_ret(pid, -errno.EPERM)

                if not insyscall:
                    insyscall = True
                else:
                    insyscall = False

            self.assert_(os.WIFEXITED(status))
            self.assertEqual(os.WEXITSTATUS(status), 0)

    def test_07_get_arg_first(self):
        pass

    def test_08_get_arg_second(self):
        pass

    def test_09_get_arg_third(self):
        pass

    def test_10_get_arg_fourth(self):
        pass

    def test_11_get_arg_fifth(self):
        pass

    def test_12_get_arg_sixth(self):
        pass

if __name__ == '__main__':
    unittest.main()

File: ruby/TEST_syscall_linux.rb
#!/usr/bin/env ruby
# coding: utf-8
# vim: set sw=2 sts=2 et nowrap fenc=utf-8 :

require 'test/unit'

$:.insert(0, '.libs')
require 'PinkTrace'

class TestPinkSyscall < Test::Unit::TestCase
  def test_syscall_get_arg_invalid
    assert_raise PinkTrace::BitnessError do
      PinkTrace::Syscall.get_arg 0, 1, 13
    end
    assert_raise PinkTrace::IndexError do
      PinkTrace::Syscall.get_arg 0, PinkTrace::Syscall::MAX_INDEX
    end
  end

  def test_syscall_get_arg_esrch
    assert_raise Errno::ESRCH do
      PinkTrace::Syscall.get_arg 0, 1
    end
  end

  def test_syscall_get_no_esrch
    assert_raise Errno::ESRCH do
      PinkTrace::Syscall.get_no 0
    end
  end

  def test_syscall_get_ret
    assert_raise ArgumentError do
      PinkTrace::Syscall.get_ret
    end
    assert_raise ArgumentError do
      PinkTrace::Syscall.get_ret 0, 1
    end
    assert_raise TypeError do
      PinkTrace::Syscall.get_ret 'pink'
    end
  end

  def test_syscall_get_ret_esrch
    assert_raise Errno::ESRCH do
      PinkTrace::Syscall.get_ret 0
    end
  end

  def test_syscall_name_invalid
    assert_raise PinkTrace::BitnessError do
      PinkTrace::Syscall.name 0, 13
    end
  end

  def test_syscall_set_no_invalid
    assert_raise PinkTrace::BitnessError do
      PinkTrace::Syscall.set_no 0, 1, 123
    end
  end

  def test_syscall_set_no_esrch
    assert_raise Errno::ESRCH do
      PinkTrace::Syscall.set_no 0, 1
    end
  end

  def test_syscall_set_ret_esrch
    assert_raise Errno::ESRCH do
      PinkTrace::Syscall.set_ret 0, 1
    end
  end
end

# These test cases depend on generated system call names.
# Don't run them if they weren't generated.
if PinkTrace::Syscall.name 0
  class TestPinkSyscall
    def test_syscall_get_no
      pid = fork do
        PinkTrace::Trace.me
        Process.kill 'STOP', Process.pid

        Process.kill 0, Process.pid
      end
      Process.waitpid pid
      PinkTrace::Trace.setup pid, PinkTrace::Trace::OPTION_SYSGOOD

      # Loop until we get to the kill() system call as there's no guarantee that
      # other system calls won't be called beforehand.
      event = -1
      found = false
      while event != PinkTrace::Event::EVENT_EXIT_GENUINE
        PinkTrace::Trace.syscall pid
        Process.waitpid pid

        event = PinkTrace::Event.decide
        if event == PinkTrace::Event::EVENT_SYSCALL then
          scno = PinkTrace::Syscall.get_no pid
          name = PinkTrace::Syscall.name scno
          if name == 'kill' then
            found = true
            break
          end
        end
      end

      assert found, 'Failed to get system call number for kill'

      begin PinkTrace::Trace.kill pid
      rescue Errno::ESRCH ;end
    end

    def test_syscall_set_no
      pid = fork do
        PinkTrace::Trace.me
        Process.kill 'STOP', Process.pid

        Process.kill 0, Process.pid
      end
      Process.waitpid pid
      PinkTrace::Trace.setup pid, PinkTrace::Trace::OPTION_SYSGOOD

      # Loop until we get to the kill() system call as there's no guarantee that
      # other system calls won't be called beforehand.
      event = -1
      while event != PinkTrace::Event::EVENT_EXIT_GENUINE
        PinkTrace::Trace.syscall pid
        Process.waitpid pid

        event = PinkTrace::Event.decide
        if event == PinkTrace::Event::EVENT_SYSCALL then
          scno = PinkTrace::Syscall.get_no pid
          name = PinkTrace::Syscall.name scno
          if name == 'kill' then
            PinkTrace::Syscall.set_no pid, PinkTrace::Syscall::INVALID
            scno = PinkTrace::Syscall.get_no pid
            assert(scno == PinkTrace::Syscall::INVALID, "Wrong system call no, expected: #{PinkTrace::Syscall::INVALID} got: #{scno}")
            break
          end
        end
      end

      begin PinkTrace::Trace.kill pid
      rescue Errno::ESRCH ;end
    end

    def test_syscall_get_ret_success
      pid = fork do
        PinkTrace::Trace.me
        Process.kill 'STOP', Process.pid

        Process.kill 0, Process.pid
      end
      Process.waitpid pid
      PinkTrace::Trace.setup pid, PinkTrace::Trace::OPTION_SYSGOOD

      # Loop until we get to the kill() system call as there's no guarantee that
      # other system calls won't be called beforehand.
      event = -1
      insyscall = false
      while event != PinkTrace::Event::EVENT_EXIT_GENUINE
        PinkTrace::Trace.syscall pid
        Process.waitpid pid

        event = PinkTrace::Event.decide
        if event == PinkTrace::Event::EVENT_SYSCALL then
          scno = PinkTrace::Syscall.get_no pid
          name = PinkTrace::Syscall.name scno
          if insyscall and name == 'kill' then
            ret = PinkTrace::Syscall.get_ret pid
            assert(ret == 0, "Wrong system call return, expected: 0 got: #{ret}")
            break
          end
          insyscall = insyscall ? false : true
        end
      end

      begin PinkTrace::Trace.kill pid
      rescue Errno::ESRCH ;end
    end

    def test_syscall_get_ret_fail
      pid = fork do
        PinkTrace::Trace.me
        Process.kill 'STOP', Process.pid

        begin File.open ''
        rescue Errno::ENOENT ;end
      end
      Process.waitpid pid
      PinkTrace::Trace.setup pid, PinkTrace::Trace::OPTION_SYSGOOD

      # Loop until we get to the kill() system call as there's no guarantee that
      # other system calls won't be called beforehand.
      event = -1
      insyscall = false
      while event != PinkTrace::Event::EVENT_EXIT_GENUINE
        PinkTrace::Trace.syscall pid
        Process.waitpid pid

        event = PinkTrace::Event.decide
        if event == PinkTrace::Event::EVENT_SYSCALL then
          scno = PinkTrace::Syscall.get_no pid
          name = PinkTrace::Syscall.name scno
          if insyscall and name == 'open' then
            ret = PinkTrace::Syscall.get_ret pid
            assert(ret == -Errno::ENOENT::Errno, "Wrong system call return, expected: #{-Errno::ENOENT::Errno} got: #{ret}")
            break
          end
          insyscall = insyscall ? false : true
        end
      end

      begin PinkTrace::Trace.kill pid
      rescue Errno::ESRCH ;end
    end

    def test_syscall_set_ret_success
      pid = fork do
        PinkTrace::Trace.me
        Process.kill 'STOP', Process.pid

        ret = Process.kill 0, Process.pid
        exit 0 if ret == 1
        exit 1
      end
      Process.waitpid pid
      PinkTrace::Trace.setup pid, PinkTrace::Trace::OPTION_SYSGOOD

      # Loop until we get to the kill() system call as there's no guarantee that
      # other system calls won't be called beforehand.
      event = -1
      insyscall = false
      while event != PinkTrace::Event::EVENT_EXIT_GENUINE
        PinkTrace::Trace.syscall pid
        Process.waitpid pid

        event = PinkTrace::Event.decide
        if event == PinkTrace::Event::EVENT_SYSCALL then
          scno = PinkTrace::Syscall.get_no pid
          name = PinkTrace::Syscall.name scno
          if insyscall and name == 'kill' then
            PinkTrace::Syscall.set_ret pid, 1
          end
          insyscall = insyscall ? false : true
        end
      end

      assert $?.exited?, "Child hasn't exited!"
      assert($?.exitstatus == 0, "Wrong exit status, expected: 0 got: #{$?.exitstatus}")

      begin PinkTrace::Trace.kill pid
      rescue Errno::ESRCH ;end
    end

    def test_syscall_set_ret_fail
      pid = fork do
        PinkTrace::Trace.me
        Process.kill 'STOP', Process.pid

        begin
          Process.kill 0, Process.pid
        rescue Errno::EPERM
          exit 0
        end
        exit 1
      end
      Process.waitpid pid
      PinkTrace::Trace.setup pid, PinkTrace::Trace::OPTION_SYSGOOD

      # Loop until we get to the kill() system call as there's no guarantee that
      # other system calls won't be called beforehand.
      event = -1
      insyscall = false
      while event != PinkTrace::Event::EVENT_EXIT_GENUINE
        PinkTrace::Trace.syscall pid
        Process.waitpid pid

        event = PinkTrace::Event.decide
        if event == PinkTrace::Event::EVENT_SYSCALL then
          scno = PinkTrace::Syscall.get_no pid
          name = PinkTrace::Syscall.name scno
          if insyscall and name == 'kill' then
            PinkTrace::Syscall.set_ret pid, -Errno::EPERM::Errno
          end
          insyscall = insyscall ? false : true
        end
      end

      assert $?.exited?, "Child hasn't exited!"
      assert($?.exitstatus == 0, "Wrong exit status, expected: 0 got: #{$?.exitstatus}")

      begin PinkTrace::Trace.kill pid
      rescue Errno::ESRCH ;end
    end

    def test_syscall_get_arg_first
    end

    def test_syscall_get_arg_second
    end

    def test_syscall_get_arg_third
    end

    def test_syscall_get_arg_fourth
    end

    def test_syscall_get_arg_fifth
    end

    def test_syscall_get_arg_sixth
    end
  end
end
